spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RecurrentDecoder.scala

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl._
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.serializer.{ContainerSerializable, DataConverter, ModuleData, ModuleSerializer}
+import serialization.Bigdl.{AttrValue, BigDLModule}
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+/**
+ * [[RecurrentDecoder]] module is a container of rnn cells that used to make
+ * a prediction of the next timestep based on the prediction we made from
+ * the previous timestep. Input for RecurrentDecoder is dynamically composed
+ * during training. input at t(i) is output at t(i-1), input at t(0) is
+ * user input, and user input has to be batch x ???(depends on cell type)
+ * without time information.
+
+ * Different types of rnn cells can be added using add() function. Currently
+ * only support lstmpeephole, convlstm, convlstm3D cell.
+ */
+class RecurrentDecoder[T : ClassTag](outputLength: Int)
+  (implicit ev: TensorNumeric[T]) extends Recurrent[T] {
+
+  times = outputLength
+
+  /**
+   *
+   *  modules: topology (cell)
+   *
+   * The topology (or cell) will be cloned for N times w.r.t the time dimension.
+   *
+   * @param module module to be add
+   * @return this container
+   */
+  override def add(module: AbstractModule[_ <: Activity, _ <: Activity, T]):
+    RecurrentDecoder.this.type = {
+    require(module.isInstanceOf[Cell[T]],
+      "Recurrent: contained module should be Cell type")
+    topology = module.asInstanceOf[Cell[T]]
+    preTopology = null
+    topology.ignorePreTopology = true
+    topology.cell = topology.buildModel()
+    modules += topology
+    this
+  }
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    require(input.dim == 2 || input.dim == 4 || input.dim == 5,
+      "Recurrent: input should be a 2D/4D/5D Tensor, e.g [batch, nDim], " +
+        s"current input.dim = ${input.dim}")
+
+    batchSize = input.size(batchDim)
+
+    val hiddenSize = topology.hiddensShape(0)
+    val outputSize = input.size()
+    outputSize(1) = hiddenSize
+    require(hiddenSize == input.size()(1), "hiddenSize is " +
```

Comment:
for recurrentdecoder, input is batch x outputsize x xxx, there is no time dimention. the second dimention is hiddensize.